README.md
# sliding_button

A Flutter package that enables apps to use a slide button to confirm actions.

## Installation

Just follow the basic installation of any flutter plugin by adding the package to the project's pubspec.yaml.

## Using this amazing package

You can start by adding the **Sliding Button** to your code. All properties inside the constructor are optional so it's fast to test the package. I'll list the properties later.

Simply add the following code and you'll have a fully functional **Sliding Button**:

```dart
import 'package:sliding_button/sliding_button.dart';

...
SlidingButton(),
...
```

After testing the Widget you can look for the properties below that allows you to style the (almost) the hole thing:

```dart
import 'package:slide_button/slide_button.dart';

...
SlidingButton(
  // Change the HEIGHT of the button. The width is always double.infinity
  buttonHeight = 55,
  // Change the BACKGROUND COLOR of the button
  buttonColor = Colors.green,
  // Change the TEXT COLOR of the button
  buttonTextColor = Colors.white,
  // Change the TEXT of the button
  buttonText = 'Slide to confirm...',
  // Change the MARGIN between the SLIDE BUTTON and the BUTTON
  slideButtonMargin = 7.5,
  // Change the BACKGROUND COLOR of the SLIDE BUTTON
  slideButtonColor = Colors.white,
  // Change the ICON COLOR of the ICON that goes inside the SLIDE BUTTON
  slideButtonIconColor = Colors.green,
  // Change the ICON of the widget that goes inside the SLIDE BUTTON
  slideButtonIcon = Icons.chevron_right,
  // Change the SIZE of the ICON that goes inside the SLIDE BUTTON
  slideButtonIconSize = 30.0,
  // Change the RADIUS of the BUTTON and SLIDE BUTTON
  radius = 4.0,
  // The AMOUNT OF THE TOTAL WIDTH OF THE BUTTON IN % that the user need to slide so we can consider the action as completed
  successfulThreshold = 0.9,
  // The WIDGET that will be shown when the slide action is completed
  widgetWhenSlideIsCompleted,
  // A simple VoidCallback that WILL BE CALLED WHEN THE SLIDE ACTION IS COMPLETED
  onSlideSuccessCallback,
)
...
```

## Using the Sliding Button callbacks

To get notified when a slide event is completed you'll need provide a VoidCallback to the onSlideSuccessCallback property:

```
SlidingButton(
  ...
  onSlideSuccessCallback: () {
    // Put your amazing code here
  },
)
...
```

And finally, if you want to reset the state of the **Sliding Button** you need to access it's state through a GlobalKey:

```
final GlobalKey<SlidingButtonState> _slideButtonKey = GlobalKey<SlidingButtonState>();

...

SlidingButton(
  key: _slideButtonKey,
  ...
  onSlideSuccessCallback: () {
    // This will reset the button to the initial state
    _slideButtonKey.currentState.reset();
  },
)
...
```


CHANGELOG.md
## [0.0.1] - First release with Slide Button 100% functional (maybe with few bugs)
## [0.0.2] - Fixes on formatting and added an example

* First release with Slide Button 100% functional (maybe with few bugs). For more info check the README.md


example/main.dart
import 'package:flutter/material.dart';
import 'package:slide_button/sliding_button.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {

  final GlobalKey<SlideButtonState> _slideButtonKey = GlobalKey<SlideButtonState>();

  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      // This call to setState tells the Flutter framework that something has
      // changed in this State, which causes it to rerun the build method below
      // so that the display can reflect the updated values. If we changed
      // _counter without calling setState(), then the build method would not be
      // called again, and so nothing would appear to happen.
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Center(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: Column(
          // Column is also a layout widget. It takes a list of children and
          // arranges them vertically. By default, it sizes itself to fit its
          // children horizontally, and tries to be as tall as its parent.
          //
          // Invoke "debug painting" (press "p" in the console, choose the
          // "Toggle Debug Paint" action from the Flutter Inspector in Android
          // Studio, or the "Toggle Debug Paint" command in Visual Studio Code)
          // to see the wireframe for each widget.
          //
          // Column has various properties to control how it sizes itself and
          // how it positions its children. Here we use mainAxisAlignment to
          // center the children vertically; the main axis here is the vertical
          // axis because Columns are vertical (the cross axis would be
          // horizontal).
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.display1,
            ),
            SlidingButton(
              key: _slideButtonKey,
              buttonHeight: 50,
              buttonColor: Color.fromRGBO(24, 190, 181, 1),
              slideButtonIconColor: Color.fromRGBO(24, 190, 181, 1),
              radius: 8,
              onSlideSuccessCallback: () {
                _incrementCounter();
                Future.delayed(Duration(seconds: 3), () {
                  _slideButtonKey.currentState.reset();
                });
              },
            ),
          ],
        ),
      ),
    );
  }
}


lib/slide_button.dart
library slide_button;

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:slide_button/widgets/platform_progress_indicator.dart';

class SlideButton extends StatefulWidget {

  // Change the HEIGHT of the button. The width is always double.infinity
  final double buttonHeight;
  // Change the BACKGROUND COLOR of the button
  final Color buttonColor;
  // Change the TEXT COLOR of the button
  final Color buttonTextColor;
  // Change the TEXT of the button
  final String buttonText;
  // Change the MARGIN between the SLIDE BUTTON and the BUTTON
  final double slideButtonMargin;
  // Change the BACKGROUND COLOR of the SLIDE BUTTON
  final Color slideButtonColor;
  // Change the ICON COLOR of the ICON that goes inside the SLIDE BUTTON
  final Color slideButtonIconColor;
  // Change the ICON of the widget that goes inside the SLIDE BUTTON
  final IconData slideButtonIcon;
  // Change the SIZE of the ICON that goes inside the SLIDE BUTTON
  final double slideButtonIconSize;
  // Change the RADIUS of the BUTTON and SLIDE BUTTON
  final double radius;
  // The AMOUNT OF THE TOTAL WIDTH OF THE BUTTON IN % that the user need to slide so we can consider the action as completed
  final double successfulThreshold;
  // The WIDGET that will be shown when the slide action is completed
  final Widget widgetWhenSlideIsCompleted;
  // A simple VoidCallback that WILL BE CALLED WHEN THE SLIDE ACTION IS COMPLETED
  final VoidCallback onSlideSuccessCallback;

  const SlideButton({Key key,
    this.buttonHeight = 55,
    this.buttonColor = Colors.green,
    this.buttonTextColor = Colors.white,
    this.buttonText = 'Slide to confirm...',
    this.slideButtonMargin = 7.5,
    this.slideButtonColor = Colors.white,
    this.slideButtonIconColor = Colors.green,
    this.slideButtonIcon = Icons.chevron_right,
    this.slideButtonIconSize = 30.0,
    this.radius = 4.0,
    this.successfulThreshold = 0.9,
    this.widgetWhenSlideIsCompleted,
    this.onSlideSuccessCallback,
  }) : super(key: key);

  @override
  State<StatefulWidget> createState() => SlideButtonState(
      buttonHeight: this.buttonHeight, buttonColor: this.buttonColor,
      buttonText: this.buttonText, slideButtonMargin: this.slideButtonMargin,
      slideButtonColor: this.slideButtonColor, buttonTextColor: this.buttonTextColor,
      slideButtonIconColor: this.slideButtonIconColor, slideButtonIcon: this.slideButtonIcon,
      slideButtonIconSize: this.slideButtonIconSize, radius: this.radius,
      successfulThreshold: this.successfulThreshold, widgetWhenSlideIsCompleted: this.widgetWhenSlideIsCompleted,
      onSlideSuccessCallback: this.onSlideSuccessCallback
  );

}

class SlideButtonState extends State<SlideButton> {

  final _buttonKey = GlobalKey();
  final _slideButtonKey = GlobalKey();

  // Change the HEIGHT of the button. The width is always double.infinity
  double buttonHeight;
  // Change the BACKGROUND COLOR of the button
  Color buttonColor;
  // Change the TEXT COLOR of the button
  Color buttonTextColor;
  // Change the TEXT of the button
  String buttonText;
  // Change the MARGIN between the SLIDE BUTTON and the BUTTON
  double slideButtonMargin;
  // Change the BACKGROUND COLOR of the SLIDE BUTTON
  Color slideButtonColor;
  // Change the ICON COLOR of the ICON that goes inside the SLIDE BUTTON
  Color slideButtonIconColor;
  // Change the ICON of the widget that goes inside the SLIDE BUTTON
  IconData slideButtonIcon;
  // Change the SIZE of the ICON that goes inside the SLIDE BUTTON
  double slideButtonIconSize;
  // Change the RADIUS of the BUTTON and SLIDE BUTTON
  double radius;
  // The AMOUNT OF THE TOTAL WIDTH OF THE BUTTON IN % that the user need to slide so we can consider the action as completed
  double successfulThreshold;
  // The WIDGET that will be shown when the slide action is completed
  Widget widgetWhenSlideIsCompleted;
  // A simple VoidCallback that WILL BE CALLED WHEN THE SLIDE ACTION IS COMPLETED
  VoidCallback onSlideSuccessCallback;

  // Indicates if the user has tapped on the slide button and is ready to init the slide
  bool _isSlideEnabled = false;
  // Indicates that the user has began a drag gesture inside the slide button
  bool _isSlideStarted = false;
  // Indicates that a slide event was successfully completed
  bool _hasCompletedSlideWithSuccess = false;
  // The margin offset is used to calculate the new width of the slide button when a drag gesture is happening
  double _slideButtonMarginDragOffset = 0;
  // This two properties are calculated on runtime based on the buttonHeight and slideButtonMargin
  double _slideButtonSize;
  double _slideButtonMargin;

  SlideButtonState({
    this.buttonHeight,
    this.buttonColor,
    this.buttonTextColor,
    this.buttonText,
    this.slideButtonMargin,
    this.slideButtonColor,
    this.slideButtonIconColor,
    this.slideButtonIcon,
    this.slideButtonIconSize,
    this.radius,
    this.successfulThreshold,
    this.widgetWhenSlideIsCompleted,
    this.onSlideSuccessCallback,
  });

  @override
  void initState() {
    super.initState();
    // Initialize properties used on the slide button
    _slideButtonSize = buttonHeight - (slideButtonMargin * 2);
    _slideButtonMargin = slideButtonMargin;
    // Always add a default widget for slide successful event
    if (this.widgetWhenSlideIsCompleted == null) {
      this.widgetWhenSlideIsCompleted = Center(
        child: SizedBox(
          width: buttonHeight / 3, height: buttonHeight / 3,
          child: PlatformProgressIndicator(materialValueColor: AlwaysStoppedAnimation<Color>(this.slideButtonIconColor), materialStrokeWidth: 1.3,),
        ),
      );
    }
  }

  @override
  Widget build(BuildContext context) {
    return IgnorePointer(
      ignoring: _hasCompletedSlideWithSuccess,
      child: GestureDetector(
        onTapDown: (tapDetails) {
          // Check if the tap down event has occurred inside the slide button
          final RenderBox renderBox = _slideButtonKey.currentContext.findRenderObject();
          final slideButtonOffset = renderBox.localToGlobal(Offset.zero);
          // On all positions I've added the _slideButtonMargin. Basically we use the _slideButtonMargin as a invisible touchable area that triggers the slide event
          final startXPosition = slideButtonOffset.dx - _slideButtonMargin;
          final endXPosition = startXPosition + buttonHeight + _slideButtonMargin;
          final startYPosition = slideButtonOffset.dy - _slideButtonMargin;
          final endYPosition = startYPosition + buttonHeight + _slideButtonMargin;
          // We only enable the slide gesture if the tap occurs inside the slide button
          if ((tapDetails.globalPosition.dx >= startXPosition && tapDetails.globalPosition.dx <= endXPosition) &&
              (tapDetails.globalPosition.dy >= startYPosition && tapDetails.globalPosition.dy <= endYPosition)) {
            _isSlideEnabled = true;
            _slideButtonSize = buttonHeight;
            _slideButtonMargin = 0;
            setState(() {});
          } else{
            _isSlideEnabled = false;
            _isSlideStarted = false;
          }
        },
        onTapUp: (details) {
          _isSlideEnabled = false;
          _resetSlideButton();
          setState(() {});
        },
        onTapCancel: () {
          if (!_isSlideEnabled) {
            _isSlideEnabled = false;
            _resetSlideButton();
            setState(() {});
          }
        },
        onHorizontalDragStart: (dragDetails) {
          if (_isSlideEnabled) {
            _isSlideStarted = true;
            _slideButtonSize = buttonHeight + _slideButtonMarginDragOffset;
            _slideButtonMargin = 0;
            setState(() {});
          }
        },
        onHorizontalDragUpdate: (dragUpdateDetails) {
          if (_isSlideStarted) {
            _slideButtonMarginDragOffset += dragUpdateDetails.delta.dx;
            _slideButtonSize = buttonHeight + _slideButtonMarginDragOffset;
            _slideButtonMargin = 0;
            // Check for minimum values that must be respected. We don't animate the slide button below the minimum.
            _slideButtonMarginDragOffset = _slideButtonMarginDragOffset < 0 ? 0 : _slideButtonMarginDragOffset;
            _slideButtonSize = _slideButtonSize < buttonHeight ? buttonHeight : _slideButtonSize;
            setState(() {});
          }
        },
        onHorizontalDragCancel: () {
          _isSlideStarted = false;
          _isSlideEnabled = false;
          _resetSlideButton();
          setState(() {});
        },
        onHorizontalDragEnd: (dragDetails) {
          if (_isSlideEnabled || _isSlideStarted) {
            // Check if the slide event has reached the minimum threshold to be considered a successful slide event
            final RenderBox renderBox = _buttonKey.currentContext.findRenderObject();
            if (_slideButtonSize >= successfulThreshold * renderBox.size.width) {
              _slideButtonSize = renderBox.size.width;
              _hasCompletedSlideWithSuccess = true;
              _isSlideEnabled = false;
              _isSlideStarted = false;
              // Make sure that we've called the success callback
              onSlideSuccessCallback?.call();
            } else {
              _slideButtonMarginDragOffset = 0;
              _resetSlideButton();
            }
            setState(() {});
          }
        },
        child: Card(
          clipBehavior: Clip.antiAlias,
          color: buttonColor,
          shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(this.radius)),
          elevation: 4,
          child: Container(
            key: _buttonKey,
            width: double.infinity,
            height: buttonHeight,
            child: Stack(
              children: <Widget>[
                Align(
                  alignment: Alignment.centerLeft,
                  child: Container(
                    margin: EdgeInsets.only(left: (slideButtonMargin / 2) + buttonHeight),
                    child: Text(this.buttonText.toUpperCase(), style: TextStyle(fontSize: 12, fontWeight: FontWeight.bold, color: buttonTextColor),),
                  ),
                ),
                AnimatedContainer(
                  key: _slideButtonKey,
                  margin: EdgeInsets.only(left: _slideButtonMargin, top: _slideButtonMargin),
                  duration: Duration(milliseconds: 100),
                  width: _slideButtonSize, height: _slideButtonSize,
                  decoration: BoxDecoration(
                    borderRadius: BorderRadius.circular(this.radius),
                    color: slideButtonColor,
                    boxShadow: [
                      BoxShadow(
                        color: Colors.black12,
                        blurRadius: 5.0,
                        spreadRadius: 2.0,
                      )
                    ],
                  ),
                  child: Center(
                    child: Icon(
                      this.slideButtonIcon, color: slideButtonIconColor, size: this.slideButtonIconSize,
                    ),
                  ),
                ),
                AnimatedOpacity(
                  opacity: _hasCompletedSlideWithSuccess ? 1.0 : 0.0,
                  duration: Duration(milliseconds: 300),
                  child: Container(
                    width: double.infinity, height: buttonHeight,
                    color: slideButtonColor,
                    child: Center(
                      child: widgetWhenSlideIsCompleted,
                    ),
                  ),
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }

  void _resetSlideButton() {
    _slideButtonSize = buttonHeight - (slideButtonMargin * 2);
    _slideButtonMargin = slideButtonMargin;
  }

  void reset() {
    _resetSlideButton();
    // Reset all values to make the button at his initial state
    _slideButtonMarginDragOffset = 0;
    _hasCompletedSlideWithSuccess = false;
    _isSlideEnabled = false;
    _isSlideStarted = false;
    setState(() {});
  }

}


lib/sliding_button.dart
library slide_button;

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:sliding_button/widgets/platform_progress_indicator.dart';

class SlidingButton extends StatefulWidget {
  // Change the HEIGHT of the button. The width is always double.infinity
  final double buttonHeight;
  // Change the BACKGROUND COLOR of the button
  final Color buttonColor;
  // Change the TEXT COLOR of the button
  final Color buttonTextColor;
  // Change the TEXT of the button
  final String buttonText;
  // Change the MARGIN between the SLIDE BUTTON and the BUTTON
  final double slideButtonMargin;
  // Change the BACKGROUND COLOR of the SLIDE BUTTON
  final Color slideButtonColor;
  // Change the ICON COLOR of the ICON that goes inside the SLIDE BUTTON
  final Color slideButtonIconColor;
  // Change the ICON of the widget that goes inside the SLIDE BUTTON
  final IconData slideButtonIcon;
  // Change the SIZE of the ICON that goes inside the SLIDE BUTTON
  final double slideButtonIconSize;
  // Change the RADIUS of the BUTTON and SLIDE BUTTON
  final double radius;
  // The AMOUNT OF THE TOTAL WIDTH OF THE BUTTON IN % that the user need to slide so we can consider the action as completed
  final double successfulThreshold;
  // The WIDGET that will be shown when the slide action is completed
  final Widget widgetWhenSlideIsCompleted;
  // A simple VoidCallback that WILL BE CALLED WHEN THE SLIDE ACTION IS COMPLETED
  final VoidCallback onSlideSuccessCallback;

  const SlidingButton({
    Key key,
    this.buttonHeight = 55,
    this.buttonColor = Colors.green,
    this.buttonTextColor = Colors.white,
    this.buttonText = 'Slide to confirm...',
    this.slideButtonMargin = 7.5,
    this.slideButtonColor = Colors.white,
    this.slideButtonIconColor = Colors.green,
    this.slideButtonIcon = Icons.chevron_right,
    this.slideButtonIconSize = 30.0,
    this.radius = 4.0,
    this.successfulThreshold = 0.9,
    this.widgetWhenSlideIsCompleted,
    this.onSlideSuccessCallback,
  }) : super(key: key);

  @override
  State<StatefulWidget> createState() => SlidingButtonState(
      buttonHeight: this.buttonHeight,
      buttonColor: this.buttonColor,
      buttonText: this.buttonText,
      slideButtonMargin: this.slideButtonMargin,
      slideButtonColor: this.slideButtonColor,
      buttonTextColor: this.buttonTextColor,
      slideButtonIconColor: this.slideButtonIconColor,
      slideButtonIcon: this.slideButtonIcon,
      slideButtonIconSize: this.slideButtonIconSize,
      radius: this.radius,
      successfulThreshold: this.successfulThreshold,
      widgetWhenSlideIsCompleted: this.widgetWhenSlideIsCompleted,
      onSlideSuccessCallback: this.onSlideSuccessCallback);
}

class SlidingButtonState extends State<SlidingButton> {
  final _buttonKey = GlobalKey();
  final _slideButtonKey = GlobalKey();

  // Change the HEIGHT of the button. The width is always double.infinity
  double buttonHeight;
  // Change the BACKGROUND COLOR of the button
  Color buttonColor;
  // Change the TEXT COLOR of the button
  Color buttonTextColor;
  // Change the TEXT of the button
  String buttonText;
  // Change the MARGIN between the SLIDE BUTTON and the BUTTON
  double slideButtonMargin;
  // Change the BACKGROUND COLOR of the SLIDE BUTTON
  Color slideButtonColor;
  // Change the ICON COLOR of the ICON that goes inside the SLIDE BUTTON
  Color slideButtonIconColor;
  // Change the ICON of the widget that goes inside the SLIDE BUTTON
  IconData slideButtonIcon;
  // Change the SIZE of the ICON that goes inside the SLIDE BUTTON
  double slideButtonIconSize;
  // Change the RADIUS of the BUTTON and SLIDE BUTTON
  double radius;
  // The AMOUNT OF THE TOTAL WIDTH OF THE BUTTON IN % that the user need to slide so we can consider the action as completed
  double successfulThreshold;
  // The WIDGET that will be shown when the slide action is completed
  Widget widgetWhenSlideIsCompleted;
  // A simple VoidCallback that WILL BE CALLED WHEN THE SLIDE ACTION IS COMPLETED
  VoidCallback onSlideSuccessCallback;

  // Indicates if the user has tapped on the slide button and is ready to init the slide
  bool _isSlideEnabled = false;
  // Indicates that the user has began a drag gesture inside the slide button
  bool _isSlideStarted = false;
  // Indicates that a slide event was successfully completed
  bool _hasCompletedSlideWithSuccess = false;
  // The margin offset is used to calculate the new width of the slide button when a drag gesture is happening
  double _slideButtonMarginDragOffset = 0;
  // This two properties are calculated on runtime based on the buttonHeight and slideButtonMargin
  double _slideButtonSize;
  double _slideButtonMargin;

  SlidingButtonState({
    this.buttonHeight,
    this.buttonColor,
    this.buttonTextColor,
    this.buttonText,
    this.slideButtonMargin,
    this.slideButtonColor,
    this.slideButtonIconColor,
    this.slideButtonIcon,
    this.slideButtonIconSize,
    this.radius,
    this.successfulThreshold,
    this.widgetWhenSlideIsCompleted,
    this.onSlideSuccessCallback,
  });

  @override
  void initState() {
    super.initState();
    // Initialize properties used on the slide button
    _slideButtonSize = buttonHeight - (slideButtonMargin * 2);
    _slideButtonMargin = slideButtonMargin;
    // Always add a default widget for slide successful event
    if (this.widgetWhenSlideIsCompleted == null) {
      this.widgetWhenSlideIsCompleted = Center(
        child: SizedBox(
          width: buttonHeight / 3,
          height: buttonHeight / 3,
          child: PlatformProgressIndicator(
            materialValueColor:
                AlwaysStoppedAnimation<Color>(this.slideButtonIconColor),
            materialStrokeWidth: 1.3,
          ),
        ),
      );
    }
  }

  @override
  Widget build(BuildContext context) {
    return IgnorePointer(
      ignoring: _hasCompletedSlideWithSuccess,
      child: GestureDetector(
        onTapDown: (tapDetails) {
          // Check if the tap down event has occurred inside the slide button
          final RenderBox renderBox =
              _slideButtonKey.currentContext.findRenderObject();
          final slideButtonOffset = renderBox.localToGlobal(Offset.zero);
          // On all positions I've added the _slideButtonMargin. Basically we use the _slideButtonMargin as a invisible touchable area that triggers the slide event
          final startXPosition = slideButtonOffset.dx - _slideButtonMargin;
          final endXPosition =
              startXPosition + buttonHeight + _slideButtonMargin;
          final startYPosition = slideButtonOffset.dy - _slideButtonMargin;
          final endYPosition =
              startYPosition + buttonHeight + _slideButtonMargin;
          // We only enable the slide gesture if the tap occurs inside the slide button
          if ((tapDetails.globalPosition.dx >= startXPosition &&
                  tapDetails.globalPosition.dx <= endXPosition) &&
              (tapDetails.globalPosition.dy >= startYPosition &&
                  tapDetails.globalPosition.dy <= endYPosition)) {
            _isSlideEnabled = true;
            _slideButtonSize = buttonHeight;
            _slideButtonMargin = 0;
            setState(() {});
          } else {
            _isSlideEnabled = false;
            _isSlideStarted = false;
          }
        },
        onTapUp: (details) {
          _isSlideEnabled = false;
          _resetSlideButton();
          setState(() {});
        },
        onTapCancel: () {
          if (!_isSlideEnabled) {
            _isSlideEnabled = false;
            _resetSlideButton();
            setState(() {});
          }
        },
        onHorizontalDragStart: (dragDetails) {
          if (_isSlideEnabled) {
            _isSlideStarted = true;
            _slideButtonSize = buttonHeight + _slideButtonMarginDragOffset;
            _slideButtonMargin = 0;
            setState(() {});
          }
        },
        onHorizontalDragUpdate: (dragUpdateDetails) {
          if (_isSlideStarted) {
            _slideButtonMarginDragOffset += dragUpdateDetails.delta.dx;
            _slideButtonSize = buttonHeight + _slideButtonMarginDragOffset;
            _slideButtonMargin = 0;
            // Check for minimum values that must be respected. We don't animate the slide button below the minimum.
            _slideButtonMarginDragOffset = _slideButtonMarginDragOffset < 0
                ? 0
                : _slideButtonMarginDragOffset;
            _slideButtonSize = _slideButtonSize < buttonHeight
                ? buttonHeight
                : _slideButtonSize;
            setState(() {});
          }
        },
        onHorizontalDragCancel: () {
          _isSlideStarted = false;
          _isSlideEnabled = false;
          _resetSlideButton();
          setState(() {});
        },
        onHorizontalDragEnd: (dragDetails) {
          if (_isSlideEnabled || _isSlideStarted) {
            // Check if the slide event has reached the minimum threshold to be considered a successful slide event
            final RenderBox renderBox =
                _buttonKey.currentContext.findRenderObject();
            if (_slideButtonSize >=
                successfulThreshold * renderBox.size.width) {
              _slideButtonSize = renderBox.size.width;
              _hasCompletedSlideWithSuccess = true;
              _isSlideEnabled = false;
              _isSlideStarted = false;
              // Make sure that we've called the success callback
              onSlideSuccessCallback?.call();
            } else {
              _slideButtonMarginDragOffset = 0;
              _resetSlideButton();
            }
            setState(() {});
          }
        },
        child: Card(
          clipBehavior: Clip.antiAlias,
          color: buttonColor,
          shape: RoundedRectangleBorder(
              borderRadius: BorderRadius.circular(this.radius)),
          elevation: 4,
          child: Container(
            key: _buttonKey,
            width: double.infinity,
            height: buttonHeight,
            child: Stack(
              children: <Widget>[
                Align(
                  alignment: Alignment.centerLeft,
                  child: Container(
                    margin: EdgeInsets.only(
                        left: (slideButtonMargin / 2) + buttonHeight),
                    child: Text(
                      this.buttonText.toUpperCase(),
                      style: TextStyle(
                          fontSize: 12,
                          fontWeight: FontWeight.bold,
                          color: buttonTextColor),
                    ),
                  ),
                ),
                AnimatedContainer(
                  key: _slideButtonKey,
                  margin: EdgeInsets.only(
                      left: _slideButtonMargin, top: _slideButtonMargin),
                  duration: Duration(milliseconds: 100),
                  width: _slideButtonSize,
                  height: _slideButtonSize,
                  decoration: BoxDecoration(
                    borderRadius: BorderRadius.circular(this.radius),
                    color: slideButtonColor,
                    boxShadow: [
                      BoxShadow(
                        color: Colors.black12,
                        blurRadius: 5.0,
                        spreadRadius: 2.0,
                      )
                    ],
                  ),
                  child: Center(
                    child: Icon(
                      this.slideButtonIcon,
                      color: slideButtonIconColor,
                      size: this.slideButtonIconSize,
                    ),
                  ),
                ),
                AnimatedOpacity(
                  opacity: _hasCompletedSlideWithSuccess ? 1.0 : 0.0,
                  duration: Duration(milliseconds: 300),
                  child: Container(
                    width: double.infinity,
                    height: buttonHeight,
                    color: slideButtonColor,
                    child: Center(
                      child: widgetWhenSlideIsCompleted,
                    ),
                  ),
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }

  void _resetSlideButton() {
    _slideButtonSize = buttonHeight - (slideButtonMargin * 2);
    _slideButtonMargin = slideButtonMargin;
  }

  void reset() {
    _resetSlideButton();
    // Reset all values to make the button at his initial state
    _slideButtonMarginDragOffset = 0;
    _hasCompletedSlideWithSuccess = false;
    _isSlideEnabled = false;
    _isSlideStarted = false;
    setState(() {});
  }
}
